### ISSUE TITLE
Thoughts on rewriting polyfill

### ISSUE THREAD

**Issue by creatorrr** [creatorrr/web-streams-polyfill#6]
Forked over from [issue on whatwg repo](https://github.com/whatwg/streams/issues/630#issuecomment-267768022).

**Comment by creatorrr** [comment]
Cc- @neocov

**Comment by neovov** [comment]
*neo**v**ov ;)

**Comment by creatorrr** [comment]
Haha, my bad @neovov. Do you have any thoughts to begin with? 

I am thinking we should certainly start with:
 - Choosing target browsers and what is our compatibility target
 - Thoughts on technologies (es7, typescript?)
 - collaboration tools? (Gh issues? Slack? IRC?)
 - setting some preliminary goals 

I'm thinking of starting with a basic test suite and setting some milestones to track stuff. 

What do you think?

**Comment by neovov** [comment]
- As of target browsers what about the 2 latests versions? I'm not sure if there some required API in order to implement streams, older versions should be OK as well but I think we should avoid the clutter to test every versions
- About technology, I'd be keen to go for ES7. I'm not familiar with typescript so it's up to you.
- GH issues are mandatory 😄 , Slack could be interesting for support/brainstorming (don't know how to set ip up though)

I agree with you, we should have a strong test suite, based on what is currently used by the whatwg and build around it.

This week-end I took some time to start something using webpack, I should share it with you (we should create a new branch).

**Comment by neovov** [comment]
I have started [re-writing](https://github.com/neovov/web-streams-polyfill) the reference implementation so we can discuss.

**Comment by creatorrr** [comment]
Hey @neovov, sorry for such a terribly late reply. I broke my wrist in an accident and couldn't work much at all until now. Anyhow, I've added you as a collaborator and merged your changes into a separate branch called `rewrite`.

**Comment by creatorrr** [comment]
> - As of target browsers what about the 2 latests versions? I'm not sure if there some required API in order to implement streams, older versions should be OK as well but I think we should avoid the clutter to test every versions

Agreed, let's just focus on getting a working implementation up and running. We can worry about compatibility and support later.

> - About technology, I'd be keen to go for ES7. I'm not familiar with typescript so it's up to you.

I do think that some level of static type checking can be **really helpful**. I like typescript but going down that road can mean a steep learning curve for new contributors. Keeping that in mind, I am inclined towards using [Flow](https://flowtype.org/) because it tends to be much easier to pick up but I'm open to arguments for or against this.

> - GH issues are mandatory 😄 , Slack could be interesting for support/brainstorming (don't know how to set ip up though)

I've set up a [Gitter community](https://gitter.im/web-streams-polyfill/Lobby) for this.

**Comment by jimmywarting** [comment]
Go ES7!! 😄 
Don't like transpilers durning development - takes up so much time to compile every time you make a change... the way i do it is just with a really up to date browser (Go Chrome!)
and then transpile it before i make a release.

**Comment by creatorrr** [comment]
Sorry for the lapse on this issue. I am just recovering from a long health issue. Can we kick off things again?

Cc- @kentonv @neovov

**Comment by kentonv** [comment]
@creatorrr Sorry, why am I being CC'd?

**Comment by creatorrr** [comment]
Hey @kentonv, my bad! I think I got threads from here and @whatwg/streams

**Comment by creatorrr** [comment]
@whatwg/streams#743 mixed up. Please feel free to unsubscribe if you don't think this is relevant.

**Comment by MattiasBuelens** [comment]
It's a bit quiet around here... 😶 

Anyway, I've started working on (yet another) rewrite over at [my fork](https://github.com/MattiasBuelens/web-streams-polyfill/tree/rewrite). What I've done so far:
* I moved the reference implementation out of the `spec` submodule into the project itself, and removed that submodule. The polyfill must still pass the same web platform tests that test the reference implementation, but other than that we can change those source files to better fit the polyfill's goals.
* I switched over to ES2015 modules and replaced Browserify with Rollup. It's the future! 😄 

Next up, I want to make the polyfill play along nicely with the browser's stream implementation, so you can use it together with e.g. `Response.body` from `fetch`.
* We probably need something like `ReadableStream.fromNative(readable)` to create a polyfilled stream that wraps a native readable stream. That way, users can wrap the stream from `response.body` and use all of the newer streaming features<sup>[1]</sup> that the browser might not support yet.
* Analogously, we should have something like `ReadableStream.toNative(readable)` to create a native stream from a polyfilled one, so users can pass a stream back to another browser API. For example, to create a `new Reponse(readable)` and pass it to `FetchEvent.respondWith` inside a service worker.
* Similar ideas for `WritableStream.fromNative(writable)` and `WritableStream.toNative(writable)` to convert between native and polyfilled writable streams.
* Probably need conversions for `TransformStream` as well.

More crazy ideas:
* It would be even cooler if we can use the native implementation (if available) when appropriate. For example: browsers may optimize `readable.pipeTo(writable)`, so if we know that `readable` and `writable` are both wrappers around native streams, we could try to use the browser's implementation.

[1] For example in Edge, the native `ReadableStream` class does not have a `pipeTo` method.

**Comment by MattiasBuelens** [comment]
@JCCR I'm working on it in my free time, so progress is slow. If you're interested, feel free to follow me and/or watch my fork. 😉 

I don't think it's a good time for me to ask for help yet, since I'm still experimenting with a lot of stuff. I've had to restart the rewrite twice already because I came up with new ideas along the way. 🙃

**Comment by JCCR** [comment]
@MattiasBuelens Yes, I started looking over your branches and I thought I saw a few different starts to a new implementation. This clears things up. I will follow, and thanks for the update!

**Comment by zenmumbler** [comment]
Hey, found this discussion and figured this would be a crowd with some interest. This will be a shameless plug, but the number of people right now interested in Web Streams is not that big, so apologies :)

I implemented the whole web streams spec from scratch in TypeScript as I'm moving to use streams a lot in my own projects and I wanted to be able to edit implementation details as needed. It passes all WPT tests (except detached buffer tests since that is not implemented in user agents).
It is here: https://github.com/stardazed/sd-streams

I made a fetch adapter repo that takes any ReadableStream implementation and creates adapted fetch and Response objects to use it, similar to Mattias' code. I wrap fetch so it can accept ReadableStreams for both uploads and in the resulting Response classes for downloads.
Find this glue code generator here: https://github.com/stardazed/sd-streams-fetch-adapter

Then I combined the two and made a polyfill that patches fetch, Response and the streams for a single include make-it-work polyfill. I tested with MDN's stream examples and the polyfill allows their examples to run without modifying the code, it's completely transparent (other than the extra include in the head).
The polyfill is here: https://github.com/stardazed/sd-streams-polyfill

Hopefully this is interesting to some people here, my implementation however does not support IE, it requires a reasonably modern browser or Node and the polyfill has some more restrictions, min specs are listed in the repo. Sorry for the intrusion and thanks!
